services/configuration/src/transactions.rs

```diff
@@ -223,12 +247,18 @@ impl Transaction for Propose {
     }
 }
 
-impl Vote {
-    /// Checks context-dependent conditions for a `Vote` transaction.
+trait TxVoteUtil: Clone {
```

Comment:
IMO, it could make more sense to define an enum like this

```rust
enum VotingDecisionRef<'a> {
    Vote(&'a Vote),
    VoteAgainst(&'a VoteAgainst),
}
```

and implement methods for it:

```rust
impl<'a> VotingDecisionRef<'a> {
    fn sender(&self) -> &PublicKey {
        match *self {
            VotingDecisionRef::Vote(vote) => vote.from(),
            VotingDecisionRef::VoteAgainst(vote) => vote.from(),
        }
    }

    fn cfg_hash(&self) -> &Hash { /* ... */ }

    fn precheck(&self, snapshot: &Snapshot) -> Result<StoredConfiguration, ServiceError>
    { /* ... */ }

    fn save(&self, fork: &mut Fork) { /* ... */ }
}

impl<'a> From<&'a Vote> VotingDecisionRef<'a> { /* trivial */ }
impl<'a> From<&'a VoteAgainst> VotingDecisionRef<'a> { /* trivial */ }
```

Then it could be used in the transaction code as follows:

```rust
let vote = VotingDecisionRef::from(self);
vote.precheck(fork.as_ref()).map_err(/* ... */)?;
vote.save(fork);
// ...
```

Reply:
Agree, this is better.

Done.